### PR TITLE
nautilus: rgw: add S3 object lock feature to support object worm

### PR DESCRIPTION
**Pull request by emmericp**
backport of https://github.com/ceph/ceph/pull/26538, https://tracker.ceph.com/issues/41452

all conflicts are due to unrelated changes in adjacent code

i've removed two lines from the original diff because they seem to be an unrelated bugfix for https://github.com/ceph/ceph/pull/27993 which hasn't been backported: https://github.com/ceph/ceph/pull/26538/files#diff-bc23d02876d86236fe7928b76d2ec1cdR3445-R3446

### CODE REVIEW

**Comment by emmericp** [comment]
jenkins, render docs

**Comment by emmericp** [comment]
test failure is unrelated, is there any known problem on current nautilus?

```
171/171 Test #112: unittest_erasure_code_shec_all ..........***Timeout 3600.11 sec
```

**Comment by emmericp** [comment]
jenkins retest this please

**Comment by emmericp** [comment]
jenkins retest this please

**Comment by smithfarm** [comment]
@emmericp Needs rebase, and could you also please cherry-pick https://github.com/ceph/ceph/pull/29680 (which looks like a follow-on fix) into this PR? Thanks.

**Comment by smithfarm** [comment]
@cbodley It occurred to me that you might want to dismiss your review pending the rebase?

**Comment by emmericp** [comment]
done, conflicts were due to adjacent changes by https://github.com/ceph/ceph/pull/30579, nothing big

also cherry-picked a4128494f55b8ad3b9fb122e5d2f2064de970f8a

**Comment by smithfarm** [comment]
Build failure:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_rgw_iam_policy.cc: At global scope:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_rgw_iam_policy.cc:82:17: error: 'rgw::IAM::Service' has not been declared
 using rgw::IAM::Service;
                 ^~~~~~~
```

File: src/test/rgw/test_rgw_iam_policy.cc [hunk 1]
```diff
@@ -76,7 +76,10 @@ using rgw::IAM::s3ListMultipartUploadParts;
 using rgw::IAM::None;
 using rgw::IAM::s3PutBucketAcl;
 using rgw::IAM::s3PutBucketPolicy;
-using rgw::Service;
```
Comment (yuvalif) [inline src/test/rgw/test_rgw_iam_policy.cc]: this is a conflict issue with recent backport (https://github.com/ceph/ceph/pull/30579). please change to: ``rgw::Service``

**Comment by emmericp** [comment]
fixed the compile error in the test, missed that changed import

**Comment by emmericp** [comment]
added eb5ee0a33110dcdab6db72df48956b47c1e2ebac to get rid of a compiler warning

**Comment by yuriw** [comment]
[wip-yuri3-testing-2019-11-08-1221-nautilus](https://trello.com/c/GKBiDkwL/752-wip-yuri3-testing-2019-11-08-1221-nautilus)

**Pull request merged**